### PR TITLE
added solution - 29

### PR DESCRIPTION
**Pull request by glones**

### CODE REVIEW

**Review by SetMyDream — APPROVED** [review]
Seems Ok

File: src/main/java/com/internet/shop/models/Role.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package com.internet.shop.models;
+
+public class Role {
+    private Long id;
+    private RoleName roleName;
+
+    public Role(RoleName roleName) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/models/Role.java:7]:
```suggestion
    private Role(RoleName roleName) {
```

File: src/main/webapp/web.xml [hunk 2]
```diff
@@ -173,4 +173,22 @@
...
     </filter-mapping>
+
+    <servlet>
+        <servlet-name>logout</servlet-name>
+        <servlet-class>com.internet.shop.controllers.LogoutController</servlet-class>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>logout</servlet-name>
+        <url-pattern>/logout</url-pattern>
+    </servlet-mapping>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:184]: let's maybe place servlets first, and then filters

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AuthorizationFilter implements Filter {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java:22]: let's also refactor AuthenticationFilter as proposed in the checklist

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Override
+    public void doFilter(ServletRequest servletRequest,
+                         ServletResponse servletResponse, FilterChain filterChain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) servletRequest;
+        HttpServletResponse resp = (HttpServletResponse) servletResponse;
+        String requestedUrl = req.getServletPath();
+
+        if (protectedUrls.get(requestedUrl) == null) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java:51]: we can also use `containsKey()` here

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/models/Role.java (hunk 1 above)
Comment (glones) [reply src/main/java/com/internet/shop/models/Role.java:7]: А навіщо нам тут приватний конструктор?

File: src/main/webapp/web.xml (hunk 2 above)
Comment (glones) [reply src/main/webapp/web.xml:184]: done

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 3 above)
Comment (glones) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java:22]: done

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 4 above)
Comment (glones) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java:51]: done

File: src/main/java/com/internet/shop/models/Role.java (hunk 1 above)
Comment (KseniiaMakarova) [reply src/main/java/com/internet/shop/models/Role.java]: @glones потому что мы создаем объекты не через него, а через статический метод

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 5]
```diff
@@ -47,14 +47,10 @@ public void doFilter(ServletRequest servletRequest,
...
-
-        if (protectedUrls.get(requestedUrl) == null) {
-            filterChain.doFilter(req, resp);
-            return;
-        }
         Long userId = (Long) req.getSession().getAttribute(USER_ID);
         User user = userService.get(userId);
-        if (isAuthorized(user, protectedUrls.get(requestedUrl))) {
+        if (isAuthorized(user, protectedUrls.get(requestedUrl))
+                || !protectedUrls.containsKey(requestedUrl)) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java:53]: now that you placed everything in one `if`, does your app work when you launch it and go to `/login`?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 5 above)
Comment (glones) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: Understand, first we should create user and only than chek his role

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**